src/proc/formats-converter.cpp

```diff
@@ -0,0 +1,396 @@
+    // Converting to stream_profile to avoid dynamic casting to video/motion_stream_profile
+    const auto && is_duplicate_predicate = [&profile]( const std::shared_ptr< stream_profile_interface > & spi ) {
+        return to_profile( spi.get() ) == to_profile( profile.get() );
+    };
+
+    return std::any_of( begin( profiles ), end( profiles ), is_duplicate_predicate );
+}
+
+// Not passing const & because we modify target_profiles, would otherwise need to create a copy
+void formats_converter::prepare_to_convert( stream_profiles target_profiles )
+{
+    clear_active_cache();
+
+    // Add missing data to source profiles (was not available during get_all_possible_target_profiles)
+    update_source_data( target_profiles );
+
+    // Caching target profiles to set as processed frames profile before calling user callback
+    cache_target_profiles( target_profiles );
+
+    while( ! target_profiles.empty() )
+    {
+        const auto & best_match = find_pbf_matching_most_profiles( target_profiles );
+        auto & best_match_pbf = best_match.first;
+        auto & best_match_target_profiles = best_match.second;
+
+        // Mark matching profiles as handled
+        for( auto & profile : best_match_target_profiles )
+        {
+            const auto & matching_profiles_predicate = [&profile]( const std::shared_ptr<stream_profile_interface> & sp ) {
+                return to_profile( profile.get() ) == to_profile( sp.get() );
+            };
+            target_profiles.erase( std::remove_if( begin( target_profiles ), end( target_profiles ), matching_profiles_predicate ) );
+        }
+
+        // Retrieve source profile from cached map and generate the relevant processing block.
+        std::unordered_set<std::shared_ptr<stream_profile_interface>> current_resolved_reqs;
+        auto best_pb = best_match_pbf->generate();
+        for( const auto & target_profile : best_match_target_profiles )
+        {
+            auto & mapped_source_profiles = _target_to_source_profiles_map[to_profile( target_profile.get() )];
+
+            for( const auto & source_profile : mapped_source_profiles )
+            {
+                if( best_match_pbf->has_source( source_profile ) )
+                {
+                    current_resolved_reqs.insert( source_profile );
+
+                    // Caching processing blocks to set their callback during sensor::start
+                    _source_profile_to_converters[source_profile].insert( best_pb );
+                }
+            }
+        }
+        const stream_profiles & print_current_resolved_reqs = { current_resolved_reqs.begin(), current_resolved_reqs.end() };
+        LOG_INFO( "Request: " << best_match_target_profiles << "\nResolved to: " << print_current_resolved_reqs );
+    }
+}
+
+void formats_converter::update_source_data( const stream_profiles & target_profiles )
+{
+    for( auto & target_profile : target_profiles )
+    {
+        // TODO - original code only updates source[0], why?
```

Comment:
I think it's because in `init_stream_profiles` method they loop and set the same profile for each target, so every one of this targets has the same profile.
Maybe you can drop the upper target_profiles loop?
Unless you changed this function :)
```
 for (auto& profile : profiles)
     for (auto target : targets)
          _target_to_source_profiles_map[target].push_back(profile);
```

Reply:
Original code uses all profiles [here](https://github.com/IntelRealSense/librealsense/blob/master/src/sensor.cpp#L1574), maybe it doesn't matter because they later [access the map based on a stream](https://github.com/IntelRealSense/librealsense/blob/master/src/sensor.cpp#L1719) not a profile, but I am not sure. It's a small loop that won't take much resources, I prefer to leave it.